openassessment/xblock/static/js/src/lms/oa_datefactory.js

```diff
@@ -0,0 +1,90 @@
+
+/**
+ *
+ * A helper function to utilize DateUtils.
+ **/
+
+OpenAssessment.DateTimeFactory = function(element) {
+    this.element = element;
+};
+
+OpenAssessment.DateTimeFactory.prototype = {
+
+    report: function() {
+        var opFactory = this;
+        $('.ora-datetime', this.element).each(function() {
+            opFactory.generate($(this));
+        });
+    },
+
+    determineContext: function(el) {
+        var context;
+        context = {
+            datetime: el.data('datetime'),
+            timezone: el.data('timezone'),
+            language: el.data('language'),
+            format: ''
+        };
+        return context;
+    },
+
+    determineDateToken: function(el) {
+        var opFactory = this;
+        var dateToken = 'date';
+        if (opFactory.isValid(el.data('datetoken'))) {
+            dateToken = el.data('datetoken');
+        }
+        return dateToken;
+    },
+
+    generate: function(el) {
+        var opFactory = this;
+        (function(require) {
+            require([
+                'jquery',
+                'edx-ui-toolkit/js/utils/date-utils',
+                'edx-ui-toolkit/js/utils/string-utils'
+                ], function($, DateUtils, StringUtils) {
+
+                    var context;
+                    var localTimeString;
+                    var displayDatetime;
+                    var interpolateDict = {};
+
+                    if (opFactory.isValid(el.data('datetime'))) {
+                        context = opFactory.determineContext(el);
+                        if (opFactory.isValid(el.data('format'))) {
+                            context.format = DateUtils.dateFormatEnum[el.data('format')];
+                        }
+
+                        localTimeString = DateUtils.localize(context);
+
+                        interpolateDict[opFactory.determineDateToken(el)] = localTimeString;
+
+                        if (opFactory.isValid(el.data('string'))) {
+                            displayDatetime = StringUtils.interpolate(
+                                el.data('string'),
+                                interpolateDict
+                            );
+                        } else {
+                            displayDatetime = localTimeString;
+                        }
+                    } else {
+                        displayDatetime = StringUtils.interpolate(
+                            el.data('string'),
+                            interpolateDict
+                        );
+                    }
+                    el.text(displayDatetime);
+                }
+            );
+        }).call(this, require || RequireJS.require);
+    },
+
+    isValid: function(candidateVariable) {
+        return candidateVariable !== undefined &&
+            candidateVariable !== '' &&
+            candidateVariable !== 'Invalid date' &&
+            candidateVariable !== 'None';
+    }
```

Comment:
This is duplicative of code that lives in edx-platform (edx-platform/lms/static/js/dateutil_factory.js), but not of logic within the dateutils proper. The way this is structured here, it *should* be able to stand on its own. It's possible to piggyback the dateutil_factory from edx-platform, but it seemed a little unstable, and less 'self-contained'.